### DISCUSSION TITLE
Road map for V2.0 :world_map:

### DISCUSSION THREAD

**Discussion by connorferster** [robbievanleeuwen/section-properties#85] [General]
Hi @ccaprani , @Czarified, @Spectre5, @Agent6-6-6, @robbievanleeuwen; 

We are really quite close to being able to release v2.0. I want to thank everyone for the assistance/input/feedback to date. It has been really valuable.

I wanted to share with you all what I see as the v2.0 road map so we can hopefully get this done in the next five weeks (see note at end). 

Road map as follows:

## 👨‍💻 Shapely pre-processor code
Shapely pre-processor code ✔️ : I view this as being complete. There may be some small additional bugs yet to be discovered but I am confident, from @Agent6-6-6's review, that it's generally working smoothly.
* I made some changes to the `.plot_geometry()` method that adds some features similar to @Spectre5's PR but not all of them. I am going to review the changes from @Spectre5 and try and bring them in.

##  📝  Documentation (currently in progress):
* Index ✔️ (Updated logo with new colour map)
* Installation ✔️ 
* Structure of an Analysis ✔️ 
* Creating Geometries, Meshes, and Material Properties ✖️ (Written but requires some edits/fixes: docs not building quite right)
* Advanced Geometry Creation ✔️ 
* Running an Analysis ✖️ (Written but some docstring examples need to be updated)
* Viewing the Results ✖️ (Written with updated images but some image re-sizing is still required)
* Examples ✖️ (Text is updated but some images need to be updated)
* API Reference ✖️ (Auto-generated: incorrect rendering on some methods of Section class, same fix required in docstring as _Running an Analysis_, see above)
* Theoretical Background ❌ (Currently not written but @robbievanleeuwen is going to work on this so we have it for the v.2.0 release)

_Additional items_: 
* It would be good to update the docs to include #53, although I am not sure where the best place to put it is (Creating Geometries...? Structure of an Analysis? It's own page?)
* I would appreciate folks reading through the documentation and suggesting edits. I have made some edits on my own but I think folks will have some suggestions that will be useful.

## 🧰 Build tools
@Spectre5's PR contains some excellent updates to the package. The goal is to incorporate as many of them as possible into the v2.0 release, specifically:
* Switch to Github actions (also, #80)
* Add all the tools (e.g. black, isort, pytest, pre-commit, etc.)
* Add benchmarking

At minimum, the documentation needs to be completed prior to v2.0 release. I would also like to include the items from "Build tools". From there, I think we are good to release and then start making further decisions.

Thoughts?

**Comment by ccaprani** [comment]
@connorferster in terms of working through the remaining tasks; are you hoping we'd work off your fork, or shall you commit this PR into a new v2 branch of this repo that we could work off? And to avoid overlapping tasks, we could either (a) have a to-do check list in this discussion people can claim items to work on, or (b) create an issue for each to-do, that people can claim (along with tags)?

You can add modularizing the library of built-in sections and I'll take that on with some design input from everyone!

**Reply by connorferster** [reply]
Committing this PR onto a new v2 branch would be a good call. I also like (b) for tasks. Will also add the "section library" to the list.

**Comment by robbievanleeuwen** [comment]
Great work @connorferster and all that have been involved in this project!

I am currently reviewing the new code and writing up the theory docs. Bit sad to see the rainbow colours go on the plots, but agree the new scheme is more professional and useful :cry:

**Comment by connorferster** [comment]
@robbievanleeuwen, further to @ccaprani above, are you able to create a new `v2` branch that I can merge the `shapely_preprocessor` branch onto? This way we can checkout the `v2` branch, do doc updates, etc., and merge as we go. Once we we are ready to move `v2` to `master`, then it will be easier to do.

**Reply by connorferster** [reply]
Wait, why does it now say "Author" beside my name!?!? YIKES!!! 

Robbie, I think you should retain the "Author" mantle and maybe just make me a Maintainer?

**Reply by robbievanleeuwen** [reply]
@connorferster just created the v2 branch, feel free to merge your shapely branch.

I think you're author because you started this discussion thread ;)

**Reply by connorferster** [reply]
Lol, ok good :)

**Comment by connorferster** [comment]
The `shapely_preprocessor` code has now been updated to @robbievanleeuwen's `v2` branch. Folks can pull the updates and checkout the `v2` branch to work on documentation, etc.

**Reply by robbievanleeuwen** [reply]
Hi Connor,

Now that we don't have a pull request, what do you think would be the best way to incorporate code review comments through github? Do you think creating issues is the best platform for this?

Robbie

**Reply by robbievanleeuwen** [reply]
@connorferster 

With reference to the above, was just getting the setup.py updated in v2 and noticed a stale(?) reference to IPython in sections.py, I commented out this line because I wasn't entirely sure it should be removed but this doesn't feel like the best workflow as it's probably best to get you review the changes before committing to the v2 branch.

Do you think it's best to pull request any changes to the v2 branch? Keen to hear your thoughts on the best workflow here.

**Reply by connorferster** [reply]
Hi @robbievanleeuwen, I think submitting pull requests to the v2 branch will allow us to do efficient reviews. 

I think it will look like this:

1. Checkout the v2 branch
2. Then create a new branch for the fix, feature, docs that one would work on
3. Complete the work and then submit a pull request against the v2 branch
4. Code review can then happen

I think a good idea is for each pull request to focus on just one thing, e.g. if the idea is to edit one page of the docs, then create a branch and pull request for just that scope. This way it will be easier to prevent merge conflicts or fix them when they do come up.

Thanks for catching that fix: I have a terrible habit of leaving instrumentation in my code (I am sure there is a tools for that somewhere...)

We can post an issue if someone does not have the comfort level to fix a particular thing but I think that, generally speaking, it would be more efficient to just do the fix (again, keeping the scope small) and submit a pull request.

Let me know your thoughts on this.

**Reply by robbievanleeuwen** [reply]
Ok great, this sounds like a good plan! You can see that I'm only used to working with myself :sweat_smile: ...

**Reply by Spectre5** [reply]
flake8 and pylint would catch it :)

I agree that a PR to v2 branch makes the most sense

**Comment by robbievanleeuwen** [comment]
Hi all, I have enabled building v2 in the readthedocs, so if anyone wants to look at the current state of the docs for the v2 branch without building it locally you can find it here:

https://sectionproperties.readthedocs.io/en/v2/

**Reply by connorferster** [reply]
That's so cool. Did not know one could do that!

**Comment by ccaprani** [comment]
@robbievanleeuwen @connorferster  Did you guys decide on a formatting scheme yet? Are we to just do flake8 checks, or can we be more opinionated by using black formatting. This could even just be added to the CI.

**Comment by connorferster** [comment]
I am a fan of black, myself. Any objections?

**Comment by Czarified** [comment]
I'm a fan of black, although I don't have time right now to put any effort
towards v2... Having it built into CI sounds good.

On Wed, Oct 13, 2021 at 12:08 AM Connor Ferster ***@***.***>
wrote:

> I am a fan of black, myself. Any objections?
>
> Sent from my iPhone
>
> > On Oct 12, 2021, at 19:39, Colin Caprani ***@***.***> wrote:
> >
> > ﻿
> > @robbievanleeuwen @connorferster Did you guys decide on a formatting
> scheme yet? Are we to just do flake8 checks, or can we be more opinionated
> by using black formatting. This could even just be added to the CI.
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub, or unsubscribe.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/robbievanleeuwen/section-properties/discussions/85#discussioncomment-1468453>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABVBQOVYQCCTGG35B2RKKDDUGUH3FANCNFSM5EKZZA6Q>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>


-- 
-Ben

**Comment by ccaprani** [comment]
Just noting this Github Action recipe for later: https://github.com/rickstaa/action-black

**Reply by connorferster** [reply]
I have two weeks off of work coming up next week. Will be jumping back on sectionproperties next week!

**Reply by robbievanleeuwen** [reply]
I've got time over the next few weeks as well! I reckon we can v2 before 2022!

**Comment by connorferster** [comment]
@robbievanleeuwen

I have been thinking about `cross_section.py`. Would you be up for renaming this module to just `section.py`? 

After replacing all of the `...Section` classes with functions, I renamed the `CrossSection` class to just `Section` but I left the module name as `cross_section.py` thinking that I would perhaps change it to `section.py` in the future. Thoughts, feelings, reactions?

**Reply by robbievanleeuwen** [reply]
Yes sounds good, and I think with #132, we should rename `sections.py` to `geometry.py` as it no longer contains sections!
